ops/hookcmds.py

```diff
@@ -0,0 +1,995 @@
+    """Revoke access to a secret."""
+    args = ['secret-revoke']
+    if relation_id is not None:
+        args.extend(['--relation', str(relation_id)])
+    if app is not None:
+        args.extend(['--app', app])
+    if unit is not None:
+        args.extend(['--unit', unit])
+    args.append(id)
+    _run(*args)
+
+
+def secret_set(
+    id: str,
+    *,
+    content: dict[str, str] | None = None,
+    label: str | None = None,
+    description: str | None = None,
+    expire: datetime.datetime | None = None,
+    rotate: SecretRotate | None = None,
+    owner: str | None = None,
+):
+    """Update an existing secret."""
+    args = ['secret-set']
+    if label is not None:
+        args.extend(['--label', label])
+    if description is not None:
+        args.extend(['--description', description])
+    if expire is not None:
+        args.extend(['--expire', expire.isoformat()])
+    if rotate is not None:
+        args.extend(['--rotate', rotate.value])
+    if owner is not None:
+        args.extend(['--owner', owner])
+    args.append(id)
+
+    # Always use the --file argument to provide the content to avoid secret data
+    # ending up in the command line, where it may be logged and also is visible
+    # via /proc.
+    with tempfile.TemporaryDirectory() as tmp:
+        for k, v in (content or {}).items():
+            with open(f'{tmp}/{k}', mode='w', encoding='utf-8') as f:
+                f.write(v)
+            args.append(f'{k}#file={tmp}/{k}')
+        _run(*args)
+
+
+def state_delete(key: str):
+    """Delete server-side-state key value pairs."""
+    _run('state-delete', key)
+
+
+@overload
+def state_get(key: str) -> str: ...
+@overload
+def state_get(key: None) -> dict[str, str]: ...
+def state_get(key: str | None) -> dict[str, str] | str:
+    """Get server-side-state value."""
+    args = ['state-get']
+    if key is not None:
+        args.append(key)
+    result = _run(*args, return_output=True, use_json=True)
+    return cast('dict[str, str]', result) if key is None else cast('str', result)
+
+
+def state_set(data: Mapping[str, str], file: pathlib.Path | None = None):
+    """Set server-side-state values."""
+    args = ['state-set']
+    args.extend(f'{k}={v}' for k, v in data.items())
+    if file is not None:
+        args.extend(['--file', str(file)])
+    _run(*args)
+
+
+@overload
+def status_get(*, include_data: Literal[True], app: Literal[False]) -> UnitStatus: ...
+@overload
+def status_get(*, include_data: Literal[True], app: Literal[True]) -> AppStatus: ...
+@overload
+def status_get(*, include_data: Literal[False], app: bool = False) -> str: ...
+def status_get(*, include_data: bool = False, app: bool = False) -> AppStatus | UnitStatus | str:
+    """Get a status of a unit or an application."""
+    args = ['status-get', f'--application={str(app).lower()}']
+    if include_data:
+        args.append('--include-data')
```

Comment:
I forget -- what does this add again? Can we just always add it in the wrapper and drop the *include_data* arg?

Reply:
Additional information about the error if there is one (more info in [Matrix](https://matrix.to/#/!xzmWHtGpPfVCXKivIh:ubuntu.com/$Cn3gF3XGmWg6GTO-NssEzeaLHtitDy1d8VIgjojSDCw?via=ubuntu.com&via=matrix.org&via=xentonix.net)).

I do like that it simplifies the signature. From the user's perspective, it seems fine to just ignore it if you're not interested.

My understanding is that it is always pretty small, so it shouldn't make much of a performance difference if it's always requested (I haven't looked at the Juju code but I imagine that the info in it is easily accessible so there wouldn't be much of a performance impact there either).

I'll double-check the potential size and that it is trivial to get in Juju and if there are essentially no performance impacts make it always true.

Reply:
It looks like it's [this Juju code](https://github.com/juju/juju/blob/c908ede06ac705233c8cd61efb2e8059b033639e/internal/worker/uniter/runner/jujuc/status-get.go#L83) and it's already got everything it needs so there should be no noticeable performance impact. I found it a bit harder to figure out the limit of what might be in the data field (Juju uses statuses in so many places!) but it seems like it's bounded by the number of units.

Will add to our 1-1 list.